engine/e_bluefield.c

```diff
@@ -829,6 +924,728 @@ static int engine_pka_dh_finish(DH *dh)
 #if (OPENSSL_VERSION_NUMBER >= 0x10100000L)
 
 #ifndef NO_EC
+
+static int
+engine_pka_pkey_meths(ENGINE *e, EVP_PKEY_METHOD **pmeth, const int **nids, int nid)
+{
+    if (!pmeth)
+    {
+        *nids = engine_pka_pkey_meth_nids;
+        return sizeof_static_array(engine_pka_pkey_meth_nids) - 1;
+    }
+
+    if (nid == NID_X25519)
+    {
+        *pmeth = engine_pka_pmeth_X25519;
+        return 1;
+    }
+    else if (nid == NID_X448)
+    {
+        *pmeth = engine_pka_pmeth_X448;
+        return 1;
+    }
+
+    *pmeth = NULL;
+    return 0;
+}
+
+static int
+engine_pka_pkey_asn1_meths(ENGINE *e, EVP_PKEY_ASN1_METHOD **ameth,
+                           const int **nids, int nid)
+{
+    if (!ameth)
+    {
+        *nids = engine_pka_pkey_asn1_meth_nids;
+        return sizeof_static_array(engine_pka_pkey_asn1_meth_nids) - 1;
+    }
+
+    if (nid == NID_X25519)
+    {
+        *ameth = engine_pka_ameth_X25519;
+        return 1;
+    }
+    else if (nid == NID_X448)
+    {
+        *ameth = engine_pka_ameth_X448;
+        return 1;
+    }
+
+    *ameth = NULL;
+    return 0;
+}
+
+/* Montgomery curve 25519 PKEY operations */
+
+static int engine_pka_X25519_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey)
+{
+    ENGINE_PKA_KEYPAIR *kpair = NULL;
+
+    kpair = engine_pka_keypair_new(NID_X25519, 0, PKA_25519_PRIKEY_SIZE);
+    if (!kpair)
+    {
+        printf("ERROR: PKA_ENGINE: 25519: Invalid keypair\n");
+        goto err;
+    }
+
+    engine_pka_get_random_bytes(kpair->private_key.buf_ptr, PKA_25519_PRIKEY_SIZE);
```

Comment:
Q: Can this call fail (for some reason) ?

Reply:
Good catch, yes this function call can fail, I have added a check for return value.